sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java

```diff
@@ -693,9 +693,9 @@ public String toString() {
 
       @Nullable
       ValueProvider<ProjectPath> projectPath =
-          getTopicProvider() == null
+          getSubscriptionProvider() == null
```

Comment:
Does this require user to provide subscription rather than topic for read? The integration test test added here specifies topic.

Reply:
Here is what I can tell:

 - `PubsubUnboundedSource` is made so that you can customize the project for the subscription.
 - But `PubsubIO` is _not_ made to customize the project.

Before my change:

 - `PubsubIO` always uses the project off the topic if it is `read().fromTopic()` and leaves it null if it is `read().fromSubscription()`
 - `PubsubUnboundedSource` always requires a `project` if it is given a `topic` because it needs it to create the random subscription

After my change:

 - `PubsubIO` always uses the project from the subscription and leaves it null for `read().fromTopic()`
 - `PubsubUnboundedSource` never requires a project, because it defaults to getting it from PipelineOptions

So I guess actually since `PubsubIO` never actually provides a useful project, it could always be left null. Or probably better to refactor `PubsubUnboundedSource` to have two variants with some shared internals so there are zero nullable fields.

Reply:
Thanks. That makes sense. I was also wondering why `createReader()` also creates a random subscription if required (that does not sound correct). As you mentioned, may be best to always require project to be present in GcpOptions.